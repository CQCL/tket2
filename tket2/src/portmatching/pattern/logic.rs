use std::collections::BTreeSet;

use itertools::Itertools;
use portmatching::{self as pm, pattern::Satisfiable};

use crate::portmatching::{
    indexing::{HugrNodeID, HugrPortID},
    BranchClass, Constraint, HugrVariableID, MatchOp, Predicate,
};

use super::Uf;

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct PatternLogic {
    known_constraints: Uf<HugrNodeID, HugrPortID, MatchOp>,
    pattern_constraints: BTreeSet<Constraint>,
}

impl PatternLogic {
    pub fn from_constraints(pattern_constraints: BTreeSet<Constraint>) -> Self {
        Self {
            known_constraints: Uf::new(),
            pattern_constraints,
        }
    }

    /// Add constraint to the pattern's knowledge
    fn add_known_constraint(&mut self, constraint: &Constraint) -> Satisfiable {
        match constraint.predicate() {
            Predicate::IsOpEqual(op) => {
                let HugrVariableID::Op(node) = constraint.required_bindings()[0] else {
                    panic!("invalid key type");
                };
                self.known_constraints.set_op(node, op.clone())
            }
            &Predicate::IsWireSource(out_port) => {
                let HugrVariableID::Op(node) = constraint.required_bindings()[0] else {
                    panic!("invalid key type");
                };
                let (wire, is_linear) = match constraint.required_bindings()[1] {
                    HugrVariableID::CopyableWire(w) => (w, false),
                    HugrVariableID::LinearWire(w) => (w, true),
                    _ => panic!("invalid key type"),
                };
                self.known_constraints
                    .set_link(node, wire, out_port, is_linear)
            }
            &Predicate::IsWireSink(in_port) => {
                let HugrVariableID::Op(node) = constraint.required_bindings()[0] else {
                    panic!("invalid key type");
                };
                let (wire, is_linear) = match constraint.required_bindings()[1] {
                    HugrVariableID::CopyableWire(w) => (w, false),
                    HugrVariableID::LinearWire(w) => (w, true),
                    _ => panic!("invalid key type"),
                };
                self.known_constraints
                    .set_link(node, wire, in_port, is_linear)
            }
            Predicate::IsDistinctFrom { .. } => {
                let is_linear = matches!(
                    constraint.required_bindings()[0],
                    HugrVariableID::LinearWire(_)
                );
                let ports = constraint
                    .required_bindings()
                    .iter()
                    .map(|&key| {
                        let Ok(port) = key.try_into() else {
                            panic!("invalid key type");
                        };
                        port
                    })
                    .collect_vec();
                let fst_port = ports[0];

                let mut all_tautology = true;
                for &port in &ports[1..] {
                    match self
                        .known_constraints
                        .set_wires_not_equal(fst_port, port, is_linear)
                    {
                        Satisfiable::Yes(_) => all_tautology = false,
                        Satisfiable::No => return Satisfiable::No,
                        Satisfiable::Tautology => (),
                    }
                }
                if all_tautology {
                    return Satisfiable::Tautology;
                } else {
                    Satisfiable::Yes(())
                }
            }
        }
    }

    /// Simplify pattern constraints based on known constraints.
    ///
    /// Call this after modifying known constraints.
    ///
    fn update_constraints(&mut self) -> Satisfiable {
        let mut clone = self.clone();
        let mut satisfiable = true;
        self.pattern_constraints.retain(|constraint| {
            match clone.add_known_constraint(constraint) {
                Satisfiable::Yes(()) => true,
                Satisfiable::No => {
                    satisfiable = false;
                    false
                }
                Satisfiable::Tautology => false,
            }
        });

        if !satisfiable {
            return Satisfiable::No;
        }
        if self.is_satisifed() {
            return Satisfiable::Tautology;
        }

        Satisfiable::Yes(())
    }

    /// Whether all pattern constraints are satisfied
    fn is_satisifed(&self) -> bool {
        self.pattern_constraints.is_empty()
    }
}

impl pm::PatternLogic for PatternLogic {
    type Constraint = Constraint;

    type BranchClass = BranchClass;

    type Key = HugrVariableID;

    fn rank_classes(
        &self,
        known_bindings: &[Self::Key],
    ) -> impl Iterator<Item = (Self::BranchClass, portmatching::pattern::ClassRank)> {
        // Class rank * number of new bindings required for constraint
        self.pattern_constraints.iter().flat_map(|c| {
            let reqs = c.required_bindings();
            let n_new_bindings = reqs.iter().filter(|k| !known_bindings.contains(k)).count() as i32;
            c.predicate()
                .get_classes(c.required_bindings())
                .into_iter()
                .map(move |cls| (cls, cls.get_rank() * (2_f64.powi(n_new_bindings))))
        })
    }

    fn nominate(&self, cls: &Self::BranchClass) -> BTreeSet<Self::Constraint> {
        self.pattern_constraints
            .iter()
            .filter(|c| {
                let classes = c.predicate().get_classes(c.required_bindings());
                classes.iter().any(|c_cls| c_cls == cls)
            })
            .cloned()
            .collect()
    }

    fn apply_transitions(&self, transitions: &[Self::Constraint]) -> Vec<Satisfiable<Self>> {
        transitions
            .iter()
            .map(|constraint| {
                let mut clone = self.clone();
                if matches!(clone.add_known_constraint(constraint), Satisfiable::No) {
                    return Satisfiable::No;
                }
                match clone.update_constraints() {
                    Satisfiable::Yes(()) => Satisfiable::Yes(clone),
                    Satisfiable::No => Satisfiable::No,
                    Satisfiable::Tautology => Satisfiable::Tautology,
                }
            })
            .collect()
    }

    fn is_satisfiable(&self) -> portmatching::pattern::Satisfiable {
        if self.is_satisifed() {
            Satisfiable::Tautology
        } else {
            Satisfiable::Yes(())
        }
    }
}
