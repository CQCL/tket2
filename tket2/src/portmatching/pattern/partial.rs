use std::collections::BTreeSet;

use itertools::Itertools;
use portmatching::{
    self as pm,
    pattern::{ClassRank, Satisfiable},
};

use crate::portmatching::{
    branch::find_shared_class,
    indexing::HugrPortID,
    pattern::{compute_class_rank, get_distinct_from_classes},
    Constraint, ConstraintClass, HugrVariableID, Predicate,
};

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct PartialPattern {
    /// The set of constraints that must still be satisfied
    pattern_constraints: BTreeSet<Constraint>,
    /// All linear wires in the pattern
    all_linear_wires: BTreeSet<HugrPortID>,
    /// The subset of `all_linear_wires` that are known to be distinct
    known_distinct_wires: BTreeSet<HugrPortID>,
}

impl PartialPattern {
    pub fn new(
        pattern_constraints: BTreeSet<Constraint>,
        all_linear_wires: BTreeSet<HugrPortID>,
    ) -> Self {
        Self {
            pattern_constraints,
            known_distinct_wires: BTreeSet::new(),
            all_linear_wires,
        }
    }

    /// Whether all constraints of the pattern are satisfied
    fn is_satisifed(&self) -> bool {
        self.pattern_constraints.is_empty() && self.are_all_wires_known()
    }

    /// Whether all wires in the pattern are known to be pairwise distinct
    fn are_all_wires_known(&self) -> bool {
        self.all_wires()
            .all(|w| self.known_distinct_wires.contains(&w))
    }

    fn all_wires(&self) -> impl Iterator<Item = HugrPortID> + '_ {
        self.all_linear_wires
            .iter()
            .filter_map(|&v| HugrPortID::try_from(v).ok())
    }

    fn with_removed_constraint(
        &self,
        constraint: &portmatching::Constraint<HugrVariableID, Predicate>,
    ) -> Self {
        let mut ret = self.clone();
        let success = ret.pattern_constraints.remove(constraint);
        debug_assert!(success);
        ret
    }

    fn with_known_distinct_wire(&self, wire: HugrPortID) -> Self {
        let mut ret = self.clone();
        ret.known_distinct_wires.insert(wire);
        ret
    }

    fn to_satisfiable(&self) -> Satisfiable<Self> {
        self.clone().into_satisfiable()
    }

    fn into_satisfiable(self) -> Satisfiable<Self> {
        if self.is_satisifed() {
            Satisfiable::Tautology
        } else {
            Satisfiable::Yes(self)
        }
    }
}

impl pm::PartialPattern for PartialPattern {
    type Constraint = Constraint;
    type ConstraintClass = ConstraintClass;
    type Key = HugrVariableID;

    fn rank_classes(
        &self,
        known_bindings: &[Self::Key],
    ) -> impl Iterator<Item = (Self::ConstraintClass, ClassRank)> {
        // Class rank * number of new bindings required for constraint
        let pattern_classes = self.pattern_constraints.iter().flat_map(|c| {
            let reqs = c.required_bindings();
            let n_new_bindings = reqs.iter().filter(|k| !known_bindings.contains(k)).count() as i32;
            c.predicate()
                .get_classes(c.required_bindings())
                .into_iter()
                .map(move |cls| (cls, compute_class_rank(cls, n_new_bindings)))
        });
        let distinct_from_classes =
            get_distinct_from_classes(known_bindings, &self.known_distinct_wires, self.all_wires())
                .collect_vec();

        pattern_classes.chain(distinct_from_classes)
    }

    fn nominate(&self, cls: &Self::BranchClass) -> BTreeSet<Self::Constraint> {
        let mut constraints: BTreeSet<_> = self
            .pattern_constraints
            .iter()
            .filter(|c| {
                let classes = c.predicate().get_classes(c.required_bindings());
                classes.iter().any(|c_cls| c_cls == cls)
            })
            .cloned()
            .collect();

        // Nominate an IsDistinctFrom constraint if applicable
        if let &BranchClass::IsDistinctFromClass(w) = cls {
            if self.all_linear_wires.contains(&w) && !self.known_distinct_wires.contains(&w) {
                let pred = Predicate::new_is_distinct_from(self.known_distinct_wires.len());
                let all_wires = [w]
                    .into_iter()
                    .chain(self.known_distinct_wires.iter().copied());
                let args = all_wires.map(HugrVariableID::LinearWire).collect();
                let constraint = Constraint::try_new(pred, args).unwrap();
                constraints.insert(constraint);
            }
        }

        constraints
    }

    fn apply_transitions(&self, transitions: &[Self::Constraint]) -> Vec<Satisfiable<Self>> {
        let class = find_shared_class(transitions)
            .expect("transitions applied simultaneously must share a common class");

        use ConstraintClass::*;
        match class {
            IsOpEqualClass(_)
            | OccupyOutgoingPortClass(_, _)
            | OccupyIncomingPortClass(_, _)
            | IsWireSourceClass(_)
            | IsLinearWireSinkClass(_) => {
                let constraint_pos = transitions
                    .iter()
                    .position(|c| self.pattern_constraints.contains(c))
                    .expect("no nominated transition found");
                let constraint = &transitions[constraint_pos];
                let new_pattern = self.with_removed_constraint(constraint);
                one_hot_vec(
                    constraint_pos,
                    transitions.len(),
                    new_pattern.into_satisfiable(),
                )
            }
            IsDistinctFromClass(wire) => {
                assert!(
                    !self.known_distinct_wires.contains(&wire),
                    "no transition nominated for this branch class"
                );

                // Find the transitions to apply
                let pos_transitions = transitions.iter().enumerate().filter(|(_, c)| {
                    let other_wires = BTreeSet::<HugrPortID>::from_iter(
                        c.required_bindings()[1..]
                            .iter()
                            .map(|&w| w.try_into().unwrap()),
                    );
                    self.known_distinct_wires
                        .iter()
                        .all(|w| other_wires.contains(w))
                });

                // Sanity check: at least one transition will be applied
                let mut pos_transitions = pos_transitions.peekable();
                assert!(
                    pos_transitions.peek().is_some(),
                    "no nominated transition found"
                );

                // Build the vector of new patterns
                let mut ret = vec![Satisfiable::No; transitions.len()];
                let new_pattern = self.with_known_distinct_wire(wire);
                for (pos, _) in pos_transitions {
                    ret[pos] = new_pattern.to_satisfiable();
                }
                ret
            }
        }
    }

    fn is_satisfiable(&self) -> portmatching::pattern::Satisfiable {
        if self.is_satisifed() {
            Satisfiable::Tautology
        } else {
            Satisfiable::Yes(())
        }
    }
}

fn one_hot_vec<V: Clone>(
    constraint_pos: usize,
    len: usize,
    val: Satisfiable<V>,
) -> Vec<Satisfiable<V>> {
    let mut ret = vec![Satisfiable::No; len];
    ret[constraint_pos] = val;
    ret
}
