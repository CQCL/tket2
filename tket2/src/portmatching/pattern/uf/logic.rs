use std::collections::BTreeSet;

use itertools::Itertools;
use portmatching::{
    self as pm,
    pattern::{ClassRank, Satisfiable},
};

use crate::portmatching::{
    indexing::{HugrNodeID, HugrPortID},
    pattern::{compute_class_rank, get_distinct_from_classes},
    BranchClass, Constraint, HugrVariableID, MatchOp, Predicate,
};

use super::Uf;

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct PatternLogic {
    known_constraints: Uf<HugrNodeID, HugrPortID, MatchOp>,
    pattern_constraints: BTreeSet<Constraint>,

    /// All linear wires in the pattern
    all_linear_wires: BTreeSet<HugrPortID>,
    /// The subset of `all_linear_wires` that are known to be distinct
    known_distinct_wires: BTreeSet<HugrPortID>,
}

impl PatternLogic {
    pub fn new(
        pattern_constraints: BTreeSet<Constraint>,
        all_linear_wires: BTreeSet<HugrPortID>,
    ) -> Self {
        Self {
            known_constraints: Uf::new(),
            pattern_constraints,
            known_distinct_wires: BTreeSet::new(),
            all_linear_wires,
        }
    }

    /// Add constraint to the pattern's knowledge
    fn add_known_constraint(&mut self, constraint: &Constraint) -> Satisfiable {
        match constraint.predicate() {
            Predicate::IsOpEqual(op) => {
                let HugrVariableID::Op(node) = constraint.required_bindings()[0] else {
                    panic!("invalid key type");
                };
                self.known_constraints.set_op(node, op.clone())
            }
            &Predicate::IsWireSource(out_port) => {
                let HugrVariableID::Op(node) = constraint.required_bindings()[0] else {
                    panic!("invalid key type");
                };
                let (wire, is_linear) = match constraint.required_bindings()[1] {
                    HugrVariableID::CopyableWire(w) => (w, false),
                    HugrVariableID::LinearWire(w) => (w, true),
                    _ => panic!("invalid key type"),
                };
                self.known_constraints
                    .set_link(node, wire, out_port, is_linear)
            }
            &Predicate::IsWireSink(in_port) => {
                let HugrVariableID::Op(node) = constraint.required_bindings()[0] else {
                    panic!("invalid key type");
                };
                let (wire, is_linear) = match constraint.required_bindings()[1] {
                    HugrVariableID::CopyableWire(w) => (w, false),
                    HugrVariableID::LinearWire(w) => (w, true),
                    _ => panic!("invalid key type"),
                };
                self.known_constraints
                    .set_link(node, wire, in_port, is_linear)
            }
            Predicate::IsDistinctFrom { .. } => {
                let is_linear = matches!(
                    constraint.required_bindings()[0],
                    HugrVariableID::LinearWire(_)
                );
                let ports = constraint
                    .required_bindings()
                    .iter()
                    .map(|&key| {
                        let Ok(port) = key.try_into() else {
                            panic!("invalid key type");
                        };
                        port
                    })
                    .collect_vec();
                let fst_port = ports[0];

                let mut all_tautology = true;
                for &port in &ports[1..] {
                    match self
                        .known_constraints
                        .set_wires_not_equal(fst_port, port, is_linear)
                    {
                        Satisfiable::Yes(_) => all_tautology = false,
                        Satisfiable::No => return Satisfiable::No,
                        Satisfiable::Tautology => (),
                    }
                }

                // Update known distinct wires
                if !self.known_distinct_wires.contains(&fst_port) {
                    let all_not_equal = self
                        .known_distinct_wires
                        .iter()
                        .all(|&w| self.known_constraints.check_wires_not_equal(fst_port, w));
                    if all_not_equal {
                        self.known_distinct_wires.insert(fst_port);
                    }
                }

                if all_tautology {
                    return Satisfiable::Tautology;
                } else {
                    Satisfiable::Yes(())
                }
            }
        }
    }

    /// Simplify pattern constraints based on known constraints.
    ///
    /// Call this after modifying known constraints.
    ///
    fn update_pattern_constraints(&mut self) -> Satisfiable {
        let mut clone = self.clone();
        let mut satisfiable = true;
        self.pattern_constraints.retain(|constraint| {
            match clone.add_known_constraint(constraint) {
                Satisfiable::Yes(()) => true,
                Satisfiable::No => {
                    satisfiable = false;
                    false
                }
                Satisfiable::Tautology => false,
            }
        });

        if !satisfiable {
            return Satisfiable::No;
        }
        if self.is_satisifed() {
            return Satisfiable::Tautology;
        }

        Satisfiable::Yes(())
    }

    /// Whether all constraints of the pattern are satisfied
    fn is_satisifed(&self) -> bool {
        self.pattern_constraints.is_empty() && self.are_all_wires_known()
    }

    /// Whether all wires in the pattern are known to be pairwise distinct
    fn are_all_wires_known(&self) -> bool {
        self.all_wires()
            .all(|w| self.known_distinct_wires.contains(&w))
    }

    fn all_wires(&self) -> impl Iterator<Item = HugrPortID> + '_ {
        self.all_linear_wires
            .iter()
            .filter_map(|&v| HugrPortID::try_from(v).ok())
    }

    /// A "partial distinct from" constraint is a constraint with [Predicate::IsDistinctFrom]
    /// that does not contain enough information for us to conclude that all
    /// wires we are interested in are distinct.
    ///
    /// For example, consider the following constraint:
    /// ```text
    /// IsDistinctFrom(w1, w2, w3)
    /// ```
    /// asserting that w1 != w2 and w1 != w3. If `self.known_distinct_wires`
    /// is [w2, w3, w4], then we cannot conclude that w1 is distinct from all
    /// other wires as w1 ?= w4. In this case we ignore the constraint (and
    /// count on the fact that a stricter constraint is being added, too).
    ///
    /// We ignore such constraints to avoid blowing up the state space we keep
    /// track of.
    fn is_partial_distinct_from(&self, constraint: &Constraint) -> bool {
        if !matches!(constraint.predicate(), Predicate::IsDistinctFrom { .. }) {
            return false;
        }
        let target: HugrPortID = constraint.required_bindings()[0].try_into().unwrap();
        if self.known_distinct_wires.contains(&target) {
            return false;
        }
        if !self.all_linear_wires.contains(&target) {
            return false;
        }
        let other_wires: BTreeSet<HugrPortID> = constraint.required_bindings()[1..]
            .iter()
            .map(|&w| w.try_into().unwrap())
            .collect();
        self.known_distinct_wires
            .iter()
            .any(|&w| !other_wires.contains(&w))
    }
}

impl pm::PatternLogic for PatternLogic {
    type Constraint = Constraint;

    type BranchClass = BranchClass;

    type Key = HugrVariableID;

    fn rank_classes(
        &self,
        known_bindings: &[Self::Key],
    ) -> impl Iterator<Item = (Self::BranchClass, ClassRank)> {
        // Class rank * number of new bindings required for constraint
        let pattern_classes = self.pattern_constraints.iter().flat_map(|c| {
            let reqs = c.required_bindings();
            let n_new_bindings = reqs.iter().filter(|k| !known_bindings.contains(k)).count() as i32;
            c.predicate()
                .get_classes(c.required_bindings())
                .into_iter()
                .map(move |cls| (cls, compute_class_rank(cls, n_new_bindings)))
        });
        let distinct_from_classes =
            get_distinct_from_classes(known_bindings, &self.known_distinct_wires, self.all_wires())
                .collect_vec();

        pattern_classes.chain(distinct_from_classes)
    }

    fn nominate(&self, cls: &Self::BranchClass) -> BTreeSet<Self::Constraint> {
        let mut constraints: BTreeSet<_> = self
            .pattern_constraints
            .iter()
            .filter(|c| {
                let classes = c.predicate().get_classes(c.required_bindings());
                classes.iter().any(|c_cls| c_cls == cls)
            })
            .cloned()
            .collect();

        // Nominate an IsDistinctFrom constraint if applicable
        if let &BranchClass::IsDistinctFromClass(w) = cls {
            if self.all_linear_wires.contains(&w) && !self.known_distinct_wires.contains(&w) {
                let pred = Predicate::new_is_distinct_from(self.known_distinct_wires.len());
                let all_wires = [w]
                    .into_iter()
                    .chain(self.known_distinct_wires.iter().copied());
                let args = all_wires.map(HugrVariableID::LinearWire).collect();
                let constraint = Constraint::try_new(pred, args).unwrap();
                constraints.insert(constraint);
            }
        }

        constraints
    }

    fn apply_transitions(&self, transitions: &[Self::Constraint]) -> Vec<Satisfiable<Self>> {
        let ret = transitions
            .iter()
            .map(|constraint| {
                let mut clone = self.clone();
                if matches!(clone.add_known_constraint(constraint), Satisfiable::No) {
                    return Satisfiable::No;
                }
                // Do not propagate to DistinctFrom predicates that only match partially
                if self.is_partial_distinct_from(constraint) {
                    return Satisfiable::No;
                }
                match clone.update_pattern_constraints() {
                    Satisfiable::Yes(()) => Satisfiable::Yes(clone),
                    Satisfiable::No => Satisfiable::No,
                    Satisfiable::Tautology => Satisfiable::Tautology,
                }
            })
            .collect_vec();
        let cnt = ret
            .iter()
            .filter(|sat| matches!(sat, Satisfiable::Yes(_)))
            .count();
        if cnt > 1 {
            dbg!(&transitions);
            println!("increased to {cnt}");
        }

        ret
    }

    fn is_satisfiable(&self) -> portmatching::pattern::Satisfiable {
        if self.is_satisifed() {
            Satisfiable::Tautology
        } else {
            Satisfiable::Yes(())
        }
    }
}
