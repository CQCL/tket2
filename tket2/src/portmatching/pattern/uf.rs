//! Circuit Pattern using Union-Find for pattern matching

mod data_structure;
mod logic;
mod selector;

use data_structure::Uf;
use logic::PatternLogic;

use std::collections::BTreeSet;
use std::fmt::Debug;

use hugr::HugrView;
use itertools::Itertools;
use portmatching as pm;

use super::super::indexing::{HugrNodeID, HugrPortID};
use super::super::{Constraint, HugrVariableID};
use super::{
    all_linear_wires, canonical_var_map, check_no_empty_wire, decompose_to_constraints,
    get_io_boundary, InvalidPattern,
};
use crate::Circuit;

/// A pattern that matches a circuit exactly
#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
pub struct CircuitPatternUf {
    constraints: BTreeSet<Constraint>,
    incoming_wires: Vec<HugrPortID>,
    outgoing_wires: Vec<HugrPortID>,
    nodes: Vec<HugrNodeID>,
    linear_wires: BTreeSet<HugrPortID>,
}

impl pm::Pattern for CircuitPatternUf {
    type Key = HugrVariableID;
    type Logic = PatternLogic;
    type Constraint = Constraint;

    fn required_bindings(&self) -> Vec<Self::Key> {
        // TODO: We will need the boundary wires too for replacements
        self.nodes.iter().copied().map_into().collect()
    }

    fn into_logic(self) -> Self::Logic {
        PatternLogic::new(self.constraints, self.linear_wires)
    }
}

impl super::CircuitPattern for CircuitPatternUf {
    /// Construct a pattern from a circuit.
    fn try_from_circuit(circuit: &Circuit<impl HugrView>) -> Result<Self, InvalidPattern> {
        check_no_empty_wire(circuit)?;

        // Find the best map from hugr values to variables
        let var_map = canonical_var_map(circuit)?;
        let nodes: BTreeSet<_> = circuit.commands().map(|cmd| cmd.node()).collect();

        let constraints = decompose_to_constraints(circuit, &var_map);
        let (incoming_wires, outgoing_wires) = get_io_boundary(circuit, &var_map);
        let linear_wires = all_linear_wires(circuit, &var_map);

        let nodes = nodes
            .into_iter()
            .map(|n| {
                let HugrVariableID::Op(n) = var_map[&n.into()] else {
                    panic!("Invalid key type");
                };
                n
            })
            .collect();

        Ok(Self {
            constraints,
            incoming_wires,
            outgoing_wires,
            nodes,
            linear_wires,
        })
    }

    fn constraints(&self) -> &BTreeSet<Constraint> {
        &self.constraints
    }

    fn nodes(&self) -> &[HugrNodeID] {
        &self.nodes
    }

    fn incoming_wires(&self) -> &[HugrPortID] {
        &self.incoming_wires
    }

    fn outgoing_wires(&self) -> &[HugrPortID] {
        &self.outgoing_wires
    }
}

#[cfg(test)]
mod tests {

    use cool_asserts::assert_matches;
    use hugr::builder::{DFGBuilder, Dataflow, DataflowHugr};
    use hugr::extension::prelude::qb_t;
    use hugr::types::Signature;
    use rstest::rstest;

    use crate::extension::rotation::rotation_type;
    use crate::portmatching::pattern::CircuitPattern;
    use crate::portmatching::tests::circ_with_copy;
    use crate::portmatching::Predicate;
    use crate::utils::build_simple_circuit;
    use crate::Tk2Op;

    use super::*;

    fn h_cx() -> Circuit {
        build_simple_circuit(2, |circ| {
            circ.append(Tk2Op::CX, [0, 1])?;
            circ.append(Tk2Op::H, [0])?;
            Ok(())
        })
        .unwrap()
    }

    /// A circuit with two rotation gates in parallel, sharing a param
    fn circ_with_copy_disconnected() -> Circuit {
        let input_t = vec![qb_t(), qb_t(), rotation_type()];
        let output_t = vec![qb_t(), qb_t()];
        let mut h = DFGBuilder::new(Signature::new(input_t, output_t)).unwrap();

        let mut inps = h.input_wires();
        let qb1 = inps.next().unwrap();
        let qb2 = inps.next().unwrap();
        let f = inps.next().unwrap();

        let res = h.add_dataflow_op(Tk2Op::Rx, [qb1, f]).unwrap();
        let qb1 = res.outputs().next().unwrap();
        let res = h.add_dataflow_op(Tk2Op::Rx, [qb2, f]).unwrap();
        let qb2 = res.outputs().next().unwrap();

        h.finish_hugr_with_outputs([qb1, qb2]).unwrap().into()
    }

    #[test]
    fn construct_pattern() {
        let circ = h_cx();

        let p = CircuitPatternUf::try_from_circuit(&circ).unwrap();

        insta::assert_debug_snapshot!(p);
    }

    #[test]
    fn disconnected_pattern() {
        let circ = build_simple_circuit(2, |circ| {
            circ.append(Tk2Op::X, [0])?;
            circ.append(Tk2Op::T, [1])?;
            Ok(())
        })
        .unwrap();
        assert_eq!(
            CircuitPatternUf::try_from_circuit(&circ).unwrap_err(),
            InvalidPattern::NotConnected
        );
    }

    #[test]
    fn pattern_with_empty_qubit() {
        let circ = build_simple_circuit(2, |circ| {
            circ.append(Tk2Op::X, [0])?;
            Ok(())
        })
        .unwrap();
        assert_matches!(
            CircuitPatternUf::try_from_circuit(&circ).unwrap_err(),
            InvalidPattern::EmptyWire { .. }
        );
    }

    #[rstest]
    fn pattern_with_copy(circ_with_copy: Circuit) {
        let pattern = CircuitPatternUf::try_from_circuit(&circ_with_copy).unwrap();

        assert_eq!(pattern.incoming_wires.len(), 2);
        assert_eq!(pattern.outgoing_wires.len(), 1);

        let copyable_sink_wires = pattern
            .constraints
            .iter()
            .filter(|c| matches!(c.predicate(), Predicate::IsWireSink(_)))
            .filter_map(|c| {
                let HugrVariableID::CopyableWire(w) = c.required_bindings()[1] else {
                    return None;
                };
                Some(w)
            })
            .collect_vec();

        assert_eq!(copyable_sink_wires.len(), 2);
        assert_eq!(copyable_sink_wires[0], copyable_sink_wires[1]);

        insta::assert_debug_snapshot!(pattern);
    }

    #[test]
    fn pattern_with_copy_disconnected() {
        let circ = circ_with_copy_disconnected();
        assert_eq!(
            CircuitPatternUf::try_from_circuit(&circ).unwrap_err(),
            InvalidPattern::NotConnected
        );
    }
}
