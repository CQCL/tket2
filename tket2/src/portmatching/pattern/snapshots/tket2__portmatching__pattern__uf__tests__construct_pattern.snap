---
source: tket2/src/portmatching/pattern/uf.rs
expression: p
---
CircuitPatternUf {
    constraints: {
        IsOpEqual(MatchOp { op_name: "tket2.quantum.CX", .. })(Op(HugrNodeID { path_from_root: HugrPath([(Port(Incoming, 0), 0)]) })),
        IsOpEqual(MatchOp { op_name: "tket2.quantum.H", .. })(Op(HugrNodeID { path_from_root: HugrPath([]) })),
        IsWireSource(OutgoingPort(0))(Op(HugrNodeID { path_from_root: HugrPath([]) }), LinearWire(HugrPortID { node: HugrNodeID { path_from_root: HugrPath([]) }, port: Port(Outgoing, 0) })),
        IsWireSource(OutgoingPort(0))(Op(HugrNodeID { path_from_root: HugrPath([(Port(Incoming, 0), 0)]) }), LinearWire(HugrPortID { node: HugrNodeID { path_from_root: HugrPath([(Port(Incoming, 0), 0)]) }, port: Port(Outgoing, 0) })),
        IsWireSource(OutgoingPort(1))(Op(HugrNodeID { path_from_root: HugrPath([(Port(Incoming, 0), 0)]) }), LinearWire(HugrPortID { node: HugrNodeID { path_from_root: HugrPath([(Port(Incoming, 0), 0)]) }, port: Port(Outgoing, 1) })),
        IsWireSink(IncomingPort(0))(Op(HugrNodeID { path_from_root: HugrPath([]) }), LinearWire(HugrPortID { node: HugrNodeID { path_from_root: HugrPath([(Port(Incoming, 0), 0)]) }, port: Port(Outgoing, 0) })),
        IsWireSink(IncomingPort(0))(Op(HugrNodeID { path_from_root: HugrPath([(Port(Incoming, 0), 0)]) }), LinearWire(HugrPortID { node: HugrNodeID { path_from_root: HugrPath([(Port(Incoming, 0), 0)]) }, port: Port(Incoming, 0) })),
        IsWireSink(IncomingPort(1))(Op(HugrNodeID { path_from_root: HugrPath([(Port(Incoming, 0), 0)]) }), LinearWire(HugrPortID { node: HugrNodeID { path_from_root: HugrPath([(Port(Incoming, 0), 0)]) }, port: Port(Incoming, 1) })),
    },
    incoming_wires: [
        HugrPortID {
            node: HugrNodeID {
                path_from_root: HugrPath(
                    [
                        (
                            Port(
                                Incoming,
                                0,
                            ),
                            0,
                        ),
                    ],
                ),
            },
            port: Port(
                Incoming,
                0,
            ),
        },
        HugrPortID {
            node: HugrNodeID {
                path_from_root: HugrPath(
                    [
                        (
                            Port(
                                Incoming,
                                0,
                            ),
                            0,
                        ),
                    ],
                ),
            },
            port: Port(
                Incoming,
                1,
            ),
        },
    ],
    outgoing_wires: [
        HugrPortID {
            node: HugrNodeID {
                path_from_root: HugrPath(
                    [
                        (
                            Port(
                                Incoming,
                                0,
                            ),
                            0,
                        ),
                    ],
                ),
            },
            port: Port(
                Outgoing,
                1,
            ),
        },
        HugrPortID {
            node: HugrNodeID {
                path_from_root: HugrPath(
                    [],
                ),
            },
            port: Port(
                Outgoing,
                0,
            ),
        },
    ],
    nodes: [
        HugrNodeID {
            path_from_root: HugrPath(
                [
                    (
                        Port(
                            Incoming,
                            0,
                        ),
                        0,
                    ),
                ],
            ),
        },
        HugrNodeID {
            path_from_root: HugrPath(
                [],
            ),
        },
    ],
    linear_wires: {
        HugrPortID {
            node: HugrNodeID {
                path_from_root: HugrPath(
                    [],
                ),
            },
            port: Port(
                Outgoing,
                0,
            ),
        },
        HugrPortID {
            node: HugrNodeID {
                path_from_root: HugrPath(
                    [
                        (
                            Port(
                                Incoming,
                                0,
                            ),
                            0,
                        ),
                    ],
                ),
            },
            port: Port(
                Incoming,
                0,
            ),
        },
        HugrPortID {
            node: HugrNodeID {
                path_from_root: HugrPath(
                    [
                        (
                            Port(
                                Incoming,
                                0,
                            ),
                            0,
                        ),
                    ],
                ),
            },
            port: Port(
                Incoming,
                1,
            ),
        },
        HugrPortID {
            node: HugrNodeID {
                path_from_root: HugrPath(
                    [
                        (
                            Port(
                                Incoming,
                                0,
                            ),
                            0,
                        ),
                    ],
                ),
            },
            port: Port(
                Outgoing,
                0,
            ),
        },
        HugrPortID {
            node: HugrNodeID {
                path_from_root: HugrPath(
                    [
                        (
                            Port(
                                Incoming,
                                0,
                            ),
                            0,
                        ),
                    ],
                ),
            },
            port: Port(
                Outgoing,
                1,
            ),
        },
    },
}
